src/app/store/selectors/VehicleTestResultModel.selectors.ts

```diff
@@ -38,3 +40,88 @@ export const getSelectedVehicleTestResultModel = createSelector(
   selectFeature,
   (state: IVehicleTestResultModelState) => state.selectedTestResultModel
 );
+
+export const getFilteredTestTypeCategories = (testTypeCategories: TestTypeCategory[]) =>
+  createSelector(selectFeature, (state: IVehicleTestResultModelState) => {
+    let testRecord: TestResultModel;
+    let filteredCategories;
+
+    if (state.selectedTestResultModel) {
+      testRecord = state.selectedTestResultModel;
+      filteredCategories = testTypeCategories.filter((element: TestTypeCategory) => {
+        return filterCategories(element, testRecord);
+      });
+    }
+
+    filteredCategories = getTestTypeTree(filteredCategories, 0);
+
+    return filteredCategories;
+  });
+
+function getTestTypeTree(filteredCategories, level: number) {
+  return filteredCategories
+  ? filteredCategories.reduce((accumulator, testObj) => {
+    const value = testObj.nextTestTypesOrCategories;
+    const node: TreeData = {} as TreeData;
+    node.nodeName = testObj.name;
+    node.id = testObj.id;
+    node.children;
+
+    if (value != null) {
+      if (typeof value === 'object') {
+        node.children = getTestTypeTree(value, level + 1);
+      } else {
+        node.children = [];
+      }
+    }
+
+    return accumulator.concat(node);
+  }, [])
+  : console.error('Error: Sorry, no filtered categories available in State');
+}
+
+function filterCategories(category: TestTypeCategory, testRecord: TestResultModel) {
```

Comment:
Most of these functions do not seem to be covered here,
Can we please add a few unit tests to cover the logic?
It will make easier to compose them in the future if we used compose from lodash,
We could export them and then in VehicleTestResult.selectors.spec.ts we write unit tests for them.